src/Lynx/Form.purs

```diff
@@ -30,19 +33,30 @@ import Test.QuickCheck (class Arbitrary)
 import Test.QuickCheck.Arbitrary (genericArbitrary)
 import Type.Row (type (+))
 
+type Page f =
+  { name :: String
+  , tabs :: NonEmpty Array (Tab f)
+  }
+
+type Tab f =
+  { name :: String
+  , link :: String
+  , contents :: NonEmpty Array (TabContents f)
+  }
 
-type Page f = { name :: String
-              , tabs :: NonEmpty Array (Tab f)
-              }
+type TabContents f = Either (Section f) (Sequence f)
 
-type Tab f = { name :: String
-             , link :: String
-             , sections :: NonEmpty Array (Section f)
-             }
+type Section f =
+  { name :: String
+  , fields :: NonEmpty Array (Field f)
+  }
 
-type Section f = { name :: String
-                 , fields :: NonEmpty Array (Field f)
-                 }
+type Sequence f =
+  { name :: String
+    -- TODO: need to figure out what `template` should be
```

Comment:
Based on the discussion in #35, it seems like it would be somehting like:
```PureScript
type Sequence a
  = { name :: String
    , template :: NonEmpty Array (Template a)
    , sections :: Array (Section a)
    }

type Template a
  = { name :: a
    , visibility :: a
    , description :: a
    , key :: Key
    , input :: TemplateInput a
    }

data TemplateInput a
  = TemplateCurrency
    { default :: Maybe a
    , required :: a
    , placeholder :: a
    }
  | TemplateDateTime
    { default :: Maybe a
    , required :: a
    , placeholder :: a
    }
  | TemplateDropdown
    { default :: Maybe a
    , required :: a
    , options :: a
    , placeholder :: a
    }
  | TemplateText
    { default :: Maybe a
    , required :: a
    , placeholder :: a
    , maxLength :: Maybe a
    , minLength :: Maybe a
    }
  | TemplateToggle
    { default :: Maybe a
    }
  | TemplateTypeaheadSingle
    { default :: Maybe a
    , options :: a
    , resultValue :: a
    , results :: a
    , uri :: a
    }
```

That seem right?

Reply:
That looks great @joneshf-cn! I would just suggest a few adjustments to the first two types: I think `key`, `visibility`, `description` belong on `Sequence` instead of `Template` (`visibility` and `description` should also be on other layout types, but that's one of those things that have been punted, on as they don't impact MVP). Not sure we even need `Template` actually. I think we could make it `template :: Section a TemplateInput` if we assume that the extra type param somehow tells the child fields that inputs should be `TemplateInput a` instead of `Input a`.

Combined with the suggestion from my code review, we'd end up with something more like:

```purescript
type Sequence a
  = { name :: String
    , template :: Template (Section a)
    , key :: Key
    , values :: Array (Section a)
    -- we could even leave these last two out
    , visibility :: a
    , description :: a
    }
```

Maybe we'll want to work on our `Input` rows a bit, so we can have something like:

```purescript
type BaseInputRows f r =
  ( default :: Maybe f
  , errors :: Errors ValidationError
  | r
  )

data BaseInput f r
  = Currency (Record (BaseInputRows f + CurrencyRows f + r))
  | DateTime (Record (BaseInputRows f + DateTimeRows f + r))
  ...

type TemplateInput f = BaseInput f ()
type Input f = BaseInput f (value :: InputSource f)
```

Reply:
Should `Sequence` have `minLength` and `maxLength`?

Reply:
> Should `Sequence` have `minLength` and `maxLength`?

Eventually, we don't need to support it right now.

Reply:
Actually, I definitely think we should use the `BaseInputRows` idea. We want to make sure that `Input` and `TemplateInput` stay consistent with each other, this would lower the maintenance burden.

Reply:
I can factor the commonalities out once we have everything working. I'd like to do it in a separate PR though.

Reply:
That's fine, I could do it later as well. I'd also like to clean up those rows, I think we can shrink down the different rows we're combining to at most 2 per input type, but yeah, factoring this certainly isn't a showstopper.